src/ASM/LR/ASMu3D.C

```diff
@@ -374,40 +375,128 @@ void ASMu3D::closeFaces (int dir, int basis, int master)
   non-constant functions).
 */
 
-void ASMu3D::constrainFace (int dir, bool open, int dof, int code, char)
+void ASMu3D::constrainFace (int dir, bool open, int dof, int code, char basis)
 {
-  if (open)
-    std::cerr << "\nWARNING: ASMu3D::constrainFace, open boundary conditions not supported yet. Treating it as closed" << std::endl;
+  // figure out function index offset (when using multiple basis)
+  size_t ofs = 1;
+  for (int i = 1; i < basis; i++)
+    ofs += this->getNoNodes(i);
 
-  int bcode = code;
-  if (code > 0) {// Dirichlet projection will be performed
-    std::cerr << "\nWARNING: Projective (nonhomogenuous) dirichlet boundary conditions not implemented. ";
-    std::cerr << "\n         Performing variational diminishing approximation instead" << std::endl;
-    // dirich.push_back(DirichletFace(this->getBoundary(dir),dof,code));
+  // figure out what edge we are at
+  LR::parameterEdge edge;
+  switch(dir)
+  {
+  case -1: edge = LR::WEST;   break;
+  case  1: edge = LR::EAST;   break;
+  case -2: edge = LR::SOUTH;  break;
+  case  2: edge = LR::NORTH;  break;
+  case -3: edge = LR::BOTTOM; break;
+  case  3: edge = LR::TOP;    break;
+  default: return;
   }
-  else if (code < 0)
-    bcode = -code;
 
-  // get all the boundary functions from the LRspline object
-  std::vector<LR::Basisfunction*> thisEdge;
-  if (dir == -1)
-    lrspline->getEdgeFunctions(thisEdge, LR::WEST, 1);
-  else if (dir == 1)
-    lrspline->getEdgeFunctions(thisEdge, LR::EAST, 1);
-  else if (dir == -2)
-    lrspline->getEdgeFunctions(thisEdge, LR::SOUTH, 1);
-  else if (dir == 2)
-    lrspline->getEdgeFunctions(thisEdge, LR::NORTH, 1);
-  else if (dir == -3)
-    lrspline->getEdgeFunctions(thisEdge, LR::BOTTOM, 1);
-  else if (dir == 3)
-    lrspline->getEdgeFunctions(thisEdge, LR::TOP, 1);
-
-  std::cout << "\nNumber of constraints: " << thisEdge.size() << std::endl;
+  // fetch the right basis to consider
+  LR::LRSplineVolume* lr = this->getBasis(basis);
 
-  // enforce the boundary conditions
-  for(LR::Basisfunction* b  : thisEdge)
-    this->prescribe(b->getId()+1,dof,bcode);
+  // get all elements and functions on this edge
+  std::vector<LR::Basisfunction*> edgeFunctions;
+  std::vector<LR::Element*>       edgeElements;
+  lr->getEdgeFunctions(edgeFunctions,edge);
+  lr->getEdgeElements (edgeElements ,edge);
```

Comment:
yes. i guess kjetil kept the terminology from the lr surfaces for compatiblity. renamed some of the vars for slightly better clarity.